src/Repository.ts

```diff
@@ -357,54 +382,65 @@ export class MongoRepository<DOC, DTO = DOC> {
-            } catch (indexErr) {
-              if (
-                indexDefinition.overwrite &&
-                indexDefinition.options.name &&
-                indexErr.name === 'MongoError' &&
-                (indexErr.codeName === 'IndexKeySpecsConflict' || indexErr.codeName === 'IndexOptionsConflict')
-              ) {
-                // drop index and recreate
-                try {
-                  await ourCollection.dropIndex(indexDefinition.options.name);
-                  await ourCollection.createIndex(indexDefinition.fields, indexDefinition.options);
-                } catch (recreateErr) {
-                  reject(recreateErr);
-                }
-              } else {
-                reject(indexErr);
-              }
-            }
-          }
-        }
-
-        resolve(ourCollection);
-      });
+      let ourCollection;
+      const exist = (await db.listCollections({ name: this.options.name }).toArray())[0];
+      if (!exist) {
+        ourCollection = await this.createCollection(db);
+      }
+      ourCollection = db.collection(this.options.name);
+      this.indexDefinition(ourCollection, this.options.indexes);
+      resolve(ourCollection);
     });
   }
+
+  private async createCollection(db: Db): Promise<Collection<DOC> | Collection<Document>> {
+    try {
+      return await db.createCollection(this.options.name, {
+        size: this.options.size,
+        capped: this.options.capped,
+        max: this.options.max
+      });
+    } catch (createErr) {
+      if (createErr.codeName === 'NamespaceExists') {
+        return this.getCollection();
+      }
+      throw createErr;
+    }
+  }
+
+  private async indexDefinition(ourCollection: Collection<Document>, indexesOptions: IndexDefinition[]): Promise<void> {
+    let indexDefinition;
+    try {
+      if (indexesOptions?.length > 0) {
+        indexDefinition = indexesOptions.shift();
+        await ourCollection.createIndex(indexDefinition.fields, indexDefinition.options);
+        return this.indexDefinition(ourCollection, indexesOptions);
+      }
+      return;
+    } catch (indexErr) {
+      this.reTryIndexCreation(indexDefinition, ourCollection, indexesOptions, indexErr);
+    }
+  }
+
+  private async reTryIndexCreation(
+    indexDefinition: IndexDefinition,
+    ourCollection: Collection<Document>,
+    indexesOptions: IndexDefinition[],
+    indexErr: any
+  ): Promise<void> {
+    if (
+      indexDefinition?.overwrite &&
+      indexDefinition?.options?.name &&
+      indexErr?.name === 'MongoError' &&
+      (indexErr?.codeName === 'IndexKeySpecsConflict' || indexErr?.codeName === 'IndexOptionsConflict')
+    ) {
+      try {
+        await ourCollection.dropIndex(indexDefinition.options.name);
+        return await this.indexDefinition(ourCollection, indexesOptions);
+      } catch (recreateErr) {
+        throw recreateErr;
+      }
+    }
+    throw indexErr;
+  }
```

Comment:
Be sure to add docblocks (at least summaries) to the functions

Reply:
Sure, I will do.